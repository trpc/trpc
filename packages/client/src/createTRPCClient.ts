/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { Unsubscribable } from '@trpc/server/observable';
import type {
  AnyProcedure,
  AnyRouter,
  inferClientTypes,
  inferProcedureInput,
  inferTransformedProcedureOutput,
  IntersectionError,
  ProcedureOptions,
  ProcedureType,
  RouterRecord,
} from '@trpc/server/unstable-core-do-not-import';
import {
  createFlatProxy,
  createRecursiveProxy,
} from '@trpc/server/unstable-core-do-not-import';
import type { CreateTRPCClientOptions } from './createTRPCUntypedClient';
import type {
  TRPCSubscriptionObserver,
  UntypedClientProperties,
} from './internals/TRPCUntypedClient';
import { TRPCUntypedClient } from './internals/TRPCUntypedClient';
import type { TRPCClientError } from './TRPCClientError';

/**
 * @public
 **/
export type inferRouterClient<TRouter extends AnyRouter> =
  DecoratedProcedureRecord<TRouter, TRouter['_def']['record']>;

type ResolverDef = {
  input: any;
  output: any;
  transformer: boolean;
  errorShape: any;
};

/** @internal */
export type Resolver<TDef extends ResolverDef> = (
  input: TDef['input'],
  opts?: ProcedureOptions,
) => Promise<TDef['output']>;

type SubscriptionResolver<TDef extends ResolverDef> = (
  input: TDef['input'],
  opts?: Partial<
    TRPCSubscriptionObserver<TDef['output'], TRPCClientError<TDef>>
  > &
    ProcedureOptions,
) => Unsubscribable;

type DecorateProcedure<
  TType extends ProcedureType,
  TDef extends ResolverDef,
> = TType extends 'query'
  ? {
      query: Resolver<TDef>;
    }
  : TType extends 'mutation'
  ? {
      mutate: Resolver<TDef>;
    }
  : TType extends 'subscription'
  ? {
      subscribe: SubscriptionResolver<TDef>;
    }
  : never;

/**
 * @internal
 */
type DecoratedProcedureRecord<
  TRouter extends AnyRouter,
  TRecord extends RouterRecord,
> = {
  [TKey in keyof TRecord]: TRecord[TKey] extends infer $Value
    ? $Value extends RouterRecord
      ? DecoratedProcedureRecord<TRouter, $Value>
      : $Value extends AnyProcedure
      ? DecorateProcedure<
          $Value['_def']['type'],
          {
            input: inferProcedureInput<$Value>;
            output: inferTransformedProcedureOutput<
              inferClientTypes<TRouter>,
              $Value
            >;
            errorShape: inferClientTypes<TRouter>['errorShape'];
            transformer: inferClientTypes<TRouter>['transformer'];
          }
        >
      : never
    : never;
};

const clientCallTypeMap: Record<
  keyof DecorateProcedure<any, any>,
  ProcedureType
> = {
  query: 'query',
  mutate: 'mutation',
  subscribe: 'subscription',
};

/** @internal */
export const clientCallTypeToProcedureType = (
  clientCallType: string,
): ProcedureType => {
  return clientCallTypeMap[clientCallType as keyof typeof clientCallTypeMap];
};

/**
 * Creates a proxy client and shows type errors if you have query names that collide with built-in properties
 */
export type CreateTRPCClient<TRouter extends AnyRouter> =
  inferRouterClient<TRouter> extends infer $Value
    ? UntypedClientProperties & keyof $Value extends never
      ? inferRouterClient<TRouter>
      : IntersectionError<UntypedClientProperties & keyof $Value>
    : never;

/**
 * @internal
 */
export function createTRPCClientProxy<TRouter extends AnyRouter>(
  client: TRPCUntypedClient<TRouter>,
): CreateTRPCClient<TRouter> {
  const proxy = createRecursiveProxy<CreateTRPCClient<TRouter>>(
    ({ path, args }) => {
      const pathCopy = [...path];
      const procedureType = clientCallTypeToProcedureType(pathCopy.pop()!);

      const fullPath = pathCopy.join('.');

      return (client as any)[procedureType](fullPath, ...args);
    },
  );
  return createFlatProxy<CreateTRPCClient<TRouter>>((key) => {
    if (client.hasOwnProperty(key)) {
      return (client as any)[key as any];
    }
    if (key === '__untypedClient') {
      return client;
    }
    return proxy[key];
  });
}

export function createTRPCClient<TRouter extends AnyRouter>(
  opts: CreateTRPCClientOptions<TRouter>,
): CreateTRPCClient<TRouter> {
  const client = new TRPCUntypedClient(opts);
  const proxy = createTRPCClientProxy<TRouter>(client);
  return proxy;
}

/**
 * Get an untyped client from a proxy client
 * @internal
 */
export function getUntypedClient<TRouter extends AnyRouter>(
  client: inferRouterClient<TRouter>,
): TRPCUntypedClient<TRouter> {
  return (client as any).__untypedClient;
}
