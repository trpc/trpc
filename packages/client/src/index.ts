// TODO: Be explicit about what we export here

export * from './createTRPCUntypedClient';
export * from './createTRPCClient';
export * from './getFetch';
export * from './TRPCClientError';
export * from './links';

export {
  /**
   * @deprecated - use `createTRPCClient` instead
   */
  createTRPCClient as createTRPCProxyClient,
  /**
   * @deprecated - use `CreateTRPCClient` instead
   */
  type CreateTRPCClient as CreateTRPCProxyClient,
  /**
   * @deprecated - use `inferRouterClient` instead
   */
  type inferRouterClient as inferRouterProxyClient,
} from './createTRPCClient';
