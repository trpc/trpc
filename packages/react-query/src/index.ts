export * from '@trpc/client';

export { getQueryKey, getMutationKey } from './internals/getQueryKey';
export {
  createTRPCReact,
  type CreateTRPCReact,
  type CreateTRPCReactBase,
} from './createTRPCReact';
export type { inferReactQueryProcedureOptions } from './utils/inferReactQueryProcedure';
export { createTRPCQueryUtils } from './createTRPCQueryUtils';
