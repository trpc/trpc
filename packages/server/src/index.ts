export * from './@trpc/server';

export type {  UnsetMarker } from './unstable-core-do-not-import/utils';