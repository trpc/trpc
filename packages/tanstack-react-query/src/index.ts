export { createTRPCContext } from './internals/Context';
export type {
  TRPCOptionsProxy,
  InferInput,
  InferOutput,
} from './internals/createOptionsProxy';
export { createTRPCOptionsProxy } from './internals/createOptionsProxy';
export { useSubscription } from './internals/subscriptionOptions';
