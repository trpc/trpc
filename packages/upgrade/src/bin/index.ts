#!/usr/bin/env node
import { basename, extname } from 'node:path';
import { parse } from '@bomb.sh/args';
import { intro, isCancel, log, multiselect, outro } from '@clack/prompts';
import { version } from '../../package.json';
import { findTRPCImportReferences, getProgram } from '../lib/ast/scanners';
import { assertCleanGitTree, filterIgnored } from '../lib/git';
import { installPackage, uninstallPackage } from '../lib/pkgmgr';

const args = parse(process.argv.slice(2), {
  default: {
    force: false,
    skipTanstackQuery: false,
    verbose: false,
  },
  alias: {
    f: 'force',
    h: 'help',
    v: 'verbose',
    q: 'skipTanstackQuery',
  },
  boolean: true,
});
if (args.verbose) process.env['VERBOSE'] = '1';

intro(`tRPC Upgrade CLI v${version}`);

if (args['help']) {
  log.info(
    `
Usage: upgrade [options]

Options:
  -f, --force             Skip git status check, use with caution
  -q, --skipTanstackQuery Skip installing @trpc/tanstack-react-query package
  -v, --verbose           Enable verbose logging
  -h, --help              Show help
      `.trim(),
  );
  process.exit(0);
}

if (args.verbose) {
  log.info(`Running upgrade with args: ${JSON.stringify(args, null, 2)}`);
}

if (!args.force) {
  await assertCleanGitTree();
}

const transforms = await multiselect({
  message: 'Select transforms to run',
  options: [
    {
      value: require.resolve('@trpc/upgrade/transforms/hooksToOptions'),
      label: 'Migrate Hooks to xxxOptions API',
    },
    {
      value: require.resolve('@trpc/upgrade/transforms/provider'),
      label: 'Migrate context provider setup',
    },
  ],
});
if (isCancel(transforms)) process.exit(0);

// Make sure provider transform runs first if it's selected
const sortedTransforms = transforms.sort((a) =>
  a.includes('provider') ? -1 : 1,
);

const program = getProgram();
const sourceFiles = program.getSourceFiles();
const possibleReferences = findTRPCImportReferences(program);
const trpcFile = possibleReferences.mostUsed.file;
const trpcImportName = possibleReferences.importName;

const commitedFiles = await filterIgnored(sourceFiles);

for (const transform of sortedTransforms) {
  log.step(`Running transform: ${basename(transform, extname(transform))}`);
  const { run } = await import('jscodeshift/src/Runner.js');
  await run(transform, commitedFiles, {
    ...args,
    trpcFile,
    trpcImportName,
  });
  log.success(`Transform completed`);
}

if (!args.skipTanstackQuery) {
  log.info('Installing @trpc/tanstack-react-query');
  await installPackage('@trpc/tanstack-react-query');
  log.success('@trpc/tanstack-react-query installed');

  log.info('Uninstalling @trpc/react-query');
  await uninstallPackage('@trpc/react-query');
  log.success('@trpc/react-query uninstalled');
}

outro('Upgrade complete! ðŸŽ‰');
