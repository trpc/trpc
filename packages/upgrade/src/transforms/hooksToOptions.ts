/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable no-console */
import type {
  API,
  ArrowFunctionExpression,
  ASTPath,
  CallExpression,
  FileInfo,
  FunctionDeclaration,
  Identifier,
  MemberExpression,
  Options,
} from 'jscodeshift';
import { replaceMemberExpressionRootIndentifier } from '../lib/ast/modifiers';
import { findParentOfType } from '../lib/ast/walkers';

interface TransformOptions extends Options {
  trpcImportName?: string;
}

const hookToOptions = {
  useQuery: { lib: '@tanstack/react-query', fn: 'queryOptions' },
  useSuspenseQuery: { lib: '@tanstack/react-query', fn: 'queryOptions' },
  useInfiniteQuery: {
    lib: '@tanstack/react-query',
    fn: 'infiniteQueryOptions',
  },
  useSuspenseInfiniteQuery: {
    lib: '@tanstack/react-query',
    fn: 'infiniteQueryOptions',
  },
  useMutation: { lib: '@tanstack/react-query', fn: 'mutationOptions' },
  useSubscription: {
    lib: '@trpc/tanstack-react-query',
    fn: 'subscriptionOptions',
  },
} as const;

const utilMap = {
  fetch: 'fetchQuery',
  fetchInfinite: 'fetchInfiniteQuery',
  prefetch: 'prefetchQuery',
  prefetchInfinite: 'prefetchInfiniteQuery',
  ensureData: 'ensureQueryData',
  invalidate: 'invalidateQueries',
  reset: 'resetQueries',
  refetch: 'refetchQueries',
  cancel: 'cancelQuery',
  setData: 'setQueryData',
  setQueriesData: 'setQueriesData',
  setInfiniteData: 'setInfiniteQueryData',
  getData: 'getQueryData',
  getInfiniteData: 'getInfiniteQueryData',
  // setMutationDefaults: 'setMutationDefaults',
  // getMutationDefaults: 'getMutationDefaults',
  // isMutating: 'isMutating',
} as const;
type ProxyMethod = keyof typeof utilMap;

export default function transform(
  file: FileInfo,
  api: API,
  options: TransformOptions,
) {
  const { trpcImportName } = options;
  if (!trpcImportName) {
    throw new Error('trpcImportName is required');
  }

  const j = api.jscodeshift;
  const root = j(file.source);
  let dirtyFlag = false;

  // Traverse all functions, and _do stuff_
  root.find(j.FunctionDeclaration).forEach((path) => {
    replaceHooksWithOptions(path);
    removeSuspenseDestructuring(path);
    migrateUseUtils(path);
  });
  root.find(j.ArrowFunctionExpression).forEach((path) => {
    replaceHooksWithOptions(path);
    removeSuspenseDestructuring(path);
    migrateUseUtils(path);
  });

  if (dirtyFlag) {
    updateTRPCImport();
  }

  /**
   * === HELPER FUNCTIONS BELOW ===
   */

  function ensureUseTRPCCall(
    path: ASTPath<FunctionDeclaration | ArrowFunctionExpression>,
  ) {
    const variableDeclaration = j.variableDeclaration('const', [
      j.variableDeclarator(
        j.identifier(trpcImportName!),
        j.callExpression(j.identifier('useTRPC'), []),
      ),
    ]);

    if (j.FunctionDeclaration.check(path.node)) {
      path.node.body.body.unshift(variableDeclaration);
      dirtyFlag = true;
    } else if (j.BlockStatement.check(path.node.body)) {
      path.node.body.body.unshift(variableDeclaration);
      dirtyFlag = true;
    }
  }

  function updateTRPCImport() {
    const specifier = root.find(j.ImportSpecifier, {
      imported: { name: trpcImportName },
    });
    if (specifier.size() > 0) {
      specifier.replaceWith(j.importSpecifier(j.identifier('useTRPC')));
      dirtyFlag = true;
    }
  }

  function ensureImported(lib: string, specifier: string) {
    if (
      root
        .find(j.ImportDeclaration, {
          source: { value: lib },
        })
        .find(j.ImportSpecifier, { imported: { name: specifier } })
        .size() === 0
    ) {
      root
        .find(j.ImportDeclaration)
        .at(-1)
        .insertAfter(
          j.importDeclaration(
            [j.importSpecifier(j.identifier(specifier))],
            j.literal(lib),
          ),
        );
      dirtyFlag = true;
    }
  }

  function replaceHooksWithOptions(
    fnPath: ASTPath<FunctionDeclaration | ArrowFunctionExpression>,
  ) {
    // REplace proxy-hooks with useX(options())
    let hasInserted = false;
    for (const [hook, { fn, lib }] of Object.entries(hookToOptions)) {
      j(fnPath)
        .find(j.CallExpression, {
          callee: {
            property: { name: hook },
          },
        })
        .forEach((path) => {
          const memberExpr = path.node.callee as MemberExpression;
          if (
            !j.MemberExpression.check(memberExpr) ||
            !j.Identifier.check(memberExpr.property)
          ) {
            console.warn('Failed to identify hook call expression', path.node);
            return;
          }

          // Rename the hook to the options function
          memberExpr.property.name = fn;

          if (!hasInserted) {
            ensureUseTRPCCall(fnPath);
            hasInserted = true;
          }

          // Wrap it in the hook call
          j(path).replaceWith(
            j.callExpression(j.identifier(hook), [path.node]),
          );
          ensureImported(lib, hook);

          dirtyFlag = true;
        });
    }
  }

  // Migrate trpc.useUtils() to useQueryClient()
  function migrateUseUtils(
    path: ASTPath<FunctionDeclaration | ArrowFunctionExpression>,
  ) {
    j(path)
      .find(j.CallExpression, {
        callee: {
          property: {
            name: (name: string) => ['useContext', 'useUtils'].includes(name),
          },
        },
      })
      .forEach((path) => {
        if (
          j.VariableDeclarator.check(path.parentPath.node) &&
          j.Identifier.check(path.parentPath.node.id)
        ) {
          const oldIdentifier = path.parentPath.node.id as Identifier;

          // Find all the references to `utils` and replace with `queryClient[helperMap](trpc.PATH.queryFilter())`
          root
            .find(j.Identifier, { name: oldIdentifier.name })
            .forEach((path) => {
              if (j.MemberExpression.check(path.parent?.parent?.node)) {
                const callExprPath = findParentOfType<CallExpression>(
                  path.parentPath,
                  j.CallExpression,
                );
                if (!callExprPath) {
                  console.warn(
                    `Failed to walk up the tree to find utilMethod call expression, on file: ${file.path}`,
                    callExprPath,
                    { start: path.node.loc?.start, end: path.node.loc?.end },
                  );
                  return;
                }
                const callExpr = callExprPath.node;
                const memberExpr = callExpr.callee as MemberExpression;
                if (
                  !j.CallExpression.check(callExpr) ||
                  !j.MemberExpression.check(memberExpr)
                ) {
                  console.warn(
                    `Failed to walk up the tree to find utilMethod with a \`trpc.PATH.<call>\`, on file: ${file.path}`,
                    callExpr,
                    { start: path.node.loc?.start, end: path.node.loc?.end },
                  );
                  return;
                }

                if (
                  !(
                    j.MemberExpression.check(memberExpr.object) &&
                    j.Identifier.check(memberExpr.property) &&
                    memberExpr.property.name in utilMap
                  )
                ) {
                  console.warn(
                    'Failed to identify utilMethod from proxy call expression',
                    memberExpr,
                  );
                  return;
                }

                // Replace util.PATH.proxyMethod() with trpc.PATH.queryFilter()
                const proxyMethod = memberExpr.property.name as ProxyMethod;
                const replacedPath = replaceMemberExpressionRootIndentifier(
                  j,
                  memberExpr,
                  j.identifier(trpcImportName!),
                );
                if (!replacedPath) {
                  console.warn(
                    'Failed to wrap proxy call expression',
                    memberExpr,
                  );
                }
                memberExpr.property = j.identifier('queryFilter');

                // Wrap it in queryClient.utilMethod()
                callExprPath.replace(
                  j.memberExpression(
                    j.identifier('queryClient'),
                    j.callExpression(j.identifier(utilMap[proxyMethod]), [
                      callExpr,
                    ]),
                  ),
                );
              }
            });

          // Replace `const utils = trpc.useUtils()` with `const queryClient = useQueryClient()`
          j(path).replaceWith(
            j.callExpression(j.identifier('useQueryClient'), []),
          );
          path.parentPath.node.id = j.identifier('queryClient');
          ensureImported('@tanstack/react-query', 'useQueryClient');
        }

        dirtyFlag = true;
      });
  }

  function removeSuspenseDestructuring(
    path: ASTPath<FunctionDeclaration | ArrowFunctionExpression>,
  ) {
    // Remove suspense query destructuring
    j(path)
      .find(j.VariableDeclaration)
      .forEach((path) => {
        const declarator = j.VariableDeclarator.check(path.node.declarations[0])
          ? path.node.declarations[0]
          : null;

        if (
          !j.CallExpression.check(declarator?.init) ||
          !j.Identifier.check(declarator.init.callee) ||
          !['useSuspenseQuery', 'useSuspenseInfiniteQuery'].includes(
            declarator.init.callee.name,
          )
        ) {
          return;
        }

        const tuple = j.ArrayPattern.check(declarator?.id)
          ? declarator.id
          : null;
        const dataName = j.Identifier.check(tuple?.elements?.[0])
          ? tuple.elements[0].name
          : null;
        const queryName = j.Identifier.check(tuple?.elements?.[1])
          ? tuple.elements[1].name
          : null;

        if (queryName) {
          declarator.id = j.identifier(queryName);
          dirtyFlag = true;

          if (dataName) {
            j(path).insertAfter(
              j.variableDeclaration('const', [
                j.variableDeclarator(
                  j.identifier(dataName),
                  j.memberExpression(declarator.id, j.identifier('data')),
                ),
              ]),
            );
          }
        } else if (dataName) {
          // const [dataName] = ... => const { data: dataName } = ...
          declarator.id = j.objectPattern([
            j.property('init', j.identifier('data'), j.identifier(dataName)),
          ]);
          dirtyFlag = true;
        }
      });
  }

  return dirtyFlag ? root.toSource() : undefined;
}

export const parser = 'tsx';
