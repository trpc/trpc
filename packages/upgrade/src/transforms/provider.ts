import type { API, FileInfo, Options } from 'jscodeshift';

interface TransformOptions extends Options {
  trpcImportName?: string;
}

export default function transform(
  file: FileInfo,
  api: API,
  options: TransformOptions,
) {
  const { trpcImportName } = options;
  let routerName: string | undefined = undefined;

  const j = api.jscodeshift;
  const root = j(file.source);
  let dirtyFlag = false;

  // Find the variable declaration for `trpc`
  root.find(j.VariableDeclaration).forEach((path) => {
    const declaration = path.node.declarations[0];
    if (
      j.Identifier.check(declaration.id) &&
      declaration.id.name === trpcImportName
    ) {
      if (
        j.CallExpression.check(declaration.init) &&
        j.Identifier.check(declaration.init.callee) &&
        declaration.init.callee.name === 'createTRPCReact'
      ) {
        // Get router name ( TODO : should probably get this from the TS compiler along with the import path)
        routerName =
          declaration.init.original?.typeParameters?.params?.[0]?.typeName
            ?.name;

        // Replace the `createTRPCReact` call with `createTRPCContext`
        declaration.init.callee.name = 'createTRPCContext';

        // Destructure the result into `TRPCProvider` and `useTRPC`
        declaration.id = j.objectPattern([
          j.property.from({
            kind: 'init',
            key: j.identifier('TRPCProvider'),
            value: j.identifier('TRPCProvider'),
            shorthand: true,
          }),
          j.property.from({
            kind: 'init',
            key: j.identifier('useTRPC'),
            value: j.identifier('useTRPC'),
            shorthand: true,
          }),
        ]);

        dirtyFlag = true;
      }
    }
  });

  // Update the import statement if transformation was successful
  if (dirtyFlag) {
    root
      .find(j.ImportDeclaration, { source: { value: '@trpc/react-query' } })
      .forEach((path) => {
        path.node.source.value = '@trpc/tanstack-react-query';
        path.node.specifiers?.forEach((specifier) => {
          if (
            j.ImportSpecifier.check(specifier) &&
            specifier.imported.name === 'createTRPCReact'
          ) {
            specifier.imported.name = 'createTRPCContext';
          }
        });
      });
  }

  // Replace trpc.createClient with createTRPCClient<TRouter>
  root
    .find(j.CallExpression, {
      callee: {
        object: { name: trpcImportName },
        property: { name: 'createClient' },
      },
    })
    .forEach((path) => {
      path.node.callee = j.identifier('createTRPCClient');
      dirtyFlag = true;

      if (routerName) {
        (path.node as any).typeParameters = j.tsTypeParameterInstantiation([
          j.tsTypeReference(j.identifier(routerName)),
        ]);
      }
    });

  // Replace <trpc.Provider client={...} with <TRPCProvider trpcClient={...}
  root
    .find(j.JSXElement, {
      openingElement: {
        name: {
          object: { name: trpcImportName },
          property: { name: 'Provider' },
        },
      },
    })
    .forEach((path) => {
      path.node.openingElement.name = j.jsxIdentifier('TRPCProvider');
      if (path.node.closingElement) {
        path.node.closingElement.name = j.jsxIdentifier('TRPCProvider');
      }
      path.node.openingElement.attributes?.forEach((attr) => {
        if (j.JSXAttribute.check(attr) && attr.name.name === 'client') {
          attr.name.name = 'trpcClient';
        }
      });
      dirtyFlag = true;
    });

  // Update imports if transformations were applied
  if (dirtyFlag) {
    // Add createTRPCClient to the import from '@trpc/client'
    root
      .find(j.ImportDeclaration, {
        source: { value: '@trpc/client' },
      })
      .forEach((path) => {
        const createTRPCClientImport = j.importSpecifier(
          j.identifier('createTRPCClient'),
        );
        path.node.specifiers?.push(createTRPCClientImport);
      });

    // Replace trpc import with TRPCProvider
    root
      .find(j.ImportSpecifier, {
        imported: { name: trpcImportName },
      })
      .forEach((path) => {
        path.node.name = j.identifier('TRPCProvider');
      });
  }

  return dirtyFlag ? root.toSource() : undefined;
}

export const parser = 'tsx';
