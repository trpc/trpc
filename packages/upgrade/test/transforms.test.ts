import { readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { describe } from 'node:test';
import { applyTransform } from 'jscodeshift/src/testUtils';
import { format, resolveConfig } from 'prettier';
import { expect, it } from 'vitest';
import * as hooksTransform from '../src/transforms/hooksToOptions';
import * as providerTransform from '../src/transforms/provider';

const formatFile = async (path: string, source: string) => {
  const prettierConfig = await resolveConfig(path);
  return format(source, {
    ...prettierConfig,
    parser: 'typescript',
  });
};

const testFixture = async (file: string, transform: any) => {
  const source = readFileSync(file, 'utf-8');

  const transformed = applyTransform(
    transform,
    { trpcImportName: 'trpc' },
    { source },
    {},
  );
  const formatted = await formatFile(file, transformed);

  expect(formatted).toMatchSnapshot();
};

describe('hooks', () => {
  const literal = './__fixtures__/hooks'; // idk why but Vite seems to do some shit when the string is in-lined to URL
  const fixturesDir = new URL(literal, import.meta.url).pathname;

  const ONLY_RUN: string[] = [];

  const fixtures = ONLY_RUN.length ? ONLY_RUN : readdirSync(fixturesDir);
  it.each(fixtures)(
    'hooks %s',
    async (file) => await testFixture(join(fixturesDir, file), hooksTransform),
  );
});

describe('provider', () => {
  const literal = './__fixtures__/provider'; // idk why but Vite seems to do some shit when the string is in-lined to URL
  const fixturesDir = new URL(literal, import.meta.url).pathname;

  const ONLY_RUN: string[] = [];

  const fixtures = ONLY_RUN.length ? ONLY_RUN : readdirSync(fixturesDir);
  it.each(fixtures)(
    'provider %s',
    async (file) =>
      await testFixture(join(fixturesDir, file), providerTransform),
  );
});
