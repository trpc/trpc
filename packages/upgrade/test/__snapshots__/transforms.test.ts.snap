// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hooks basic.ts 1`] = `
"import { useQuery } from '@tanstack/react-query';
import { useTRPC } from './trpc';

export function Component() {
  const trpc = useTRPC();
  useQuery(trpc.post.list.queryOptions());
  useQuery(trpc.post.queryOptions());
  useQuery(trpc.a.b.c.d.queryOptions());
  // eslint-disable-next-line @typescript-eslint/dot-notation
  useQuery(trpc['post'].queryOptions());

  useQuery(trpc.post.byId.queryOptions({ id: 1 }));
  useQuery(trpc.num.queryOptions(1));

  useQuery(trpc.post.list.queryOptions(undefined, { staleTime: 1000 }));
  useQuery(trpc.post.byId.queryOptions({ id: 1 }, { staleTime: 1000 }));
}
"
`;

exports[`hooks existing-rq.ts 1`] = `""`;

exports[`hooks existing-rq-mix.ts 1`] = `
"import { useInfiniteQuery, useMutation, useQuery } from '@tanstack/react-query';
import { useTRPC } from './trpc';

export function Component() {
  const trpc = useTRPC();
  useQuery({ queryKey: ['a'], queryFn: () => 1 });
  useMutation({ mutationFn: () => 1 });

  useQuery(trpc.post.byId.queryOptions({ id: 1 }));
  useQuery(trpc.num.queryOptions(1));
  useInfiniteQuery(trpc.post.list.infiniteQueryOptions());
}
"
`;

exports[`hooks multiple-functions.ts 1`] = `
"import { useQuery } from '@tanstack/react-query';
import { useTRPC } from './trpc';

export function Component1() {
  const trpc = useTRPC();
  useQuery(trpc.post.list.queryOptions());
  useQuery(trpc.post.queryOptions());
  useQuery(trpc.a.b.c.d.queryOptions());
}

export function Component2() {
  const trpc = useTRPC();
  useQuery(trpc.post.list.queryOptions());
  useQuery(trpc.post.queryOptions());
  useQuery(trpc.a.b.c.d.queryOptions());
}
"
`;

exports[`hooks suspense-destructuring.ts 1`] = `
"import {
  useSuspenseInfiniteQuery,
  useSuspenseQuery,
} from '@tanstack/react-query';
import { useTRPC } from './trpc';

export function Component(channelId: string) {
  const trpc = useTRPC();
  const query = useSuspenseQuery(trpc.post.list.queryOptions());

  const data = query.data;

  const [a, b] = [1, 2];

  const query2 = useSuspenseInfiniteQuery(
    trpc.post.infinite.infiniteQueryOptions(
      { channelId },
      {
        getNextPageParam: (d) => d.nextCursor,
        // No need to refetch as we have a subscription
        refetchOnReconnect: false,
        refetchOnWindowFocus: false,
        refetchOnMount: false,
      },
    ),
  );

  const query3 = useSuspenseInfiniteQuery(
    trpc.post.infinite.infiniteQueryOptions(
      { channelId },
      {
        getNextPageParam: (d) => d.nextCursor,
      },
    ),
  );

  const data3 = query3.data;

  const { data: data4 } = useSuspenseInfiniteQuery(
    trpc.post.infinite.infiniteQueryOptions(
      { channelId },
      {
        getNextPageParam: (d) => d.nextCursor,
      },
    ),
  );
}
"
`;

exports[`hooks with-invalidate.ts 1`] = `
"import { useMutation, useQueryClient } from '@tanstack/react-query';
import { useTRPC } from './trpc';

export function Component() {
  const trpc = useTRPC();
  const queryClient = useQueryClient();

  const mutation = useMutation(
    trpc.post.create.mutationOptions({
      onSettled: () => {
        queryClient.invalidateQueries(trpc.post.queryFilter());
        queryClient.invalidateQueries(trpc.post.list.queryFilter());
        queryClient.invalidateQueries(
          trpc.post.x.y.z.longPropName.queryFilter(),
        );

        // eslint-disable-next-line @typescript-eslint/dot-notation
        queryClient.invalidateQueries(trpc['post'].queryFilter());
      },
    }),
  );
}
"
`;

exports[`provider basic.ts 1`] = `
"'use client';

import type { AppRouter } from '@acme/api';
import { createTRPCContext } from '@trpc/tanstack-react-query';

export const { TRPCProvider, useTRPC } = createTRPCContext<AppRouter>();
"
`;
